prusti-viper/src/encoder/mir/specifications/specs.rs

```diff
@@ -10,100 +16,169 @@ use prusti_interface::{
 };
 use rustc_hash::FxHashMap;
 use rustc_hir::def_id::DefId;
+use rustc_middle::ty::{subst::Subst, TypeFoldable};
+
+/// Defines the context for which we perform refinement.
+/// It can be thought of as the variants of [SpecQuery] for which we can perform refinement.
+#[derive(Debug)]
+struct RefinementContext<'qry, 'tcx> {
+    impl_query: &'qry SpecQuery<'tcx>,
+    trait_query: SpecQuery<'tcx>,
+}
+
+impl<'qry, 'tcx> RefinementContext<'qry, 'tcx> {
+    /// Tries to create a refinement context.
+    /// Returns None if refinement is not needed
+    fn try_from(env: &Environment<'tcx>, query: &'qry SpecQuery<'tcx>) -> Option<Self> {
+        match query {
+            SpecQuery::FunctionCallEncoding(FunctionCallEncodingQuery {
+                called_def_id: def_id,
+                call_substs: substs,
+                ..
+            })
+            | SpecQuery::FunctionDefEncoding(def_id, substs)
+            | SpecQuery::PureOrTrustedCheck(def_id, substs) => {
+                let (trait_def_id, trait_substs) = env.find_trait_method_substs(*def_id, substs)?;
+
+                // We need to subst the trait substs because they might be generic
+                // (with substitutions in the call substs)
+                let trait_substs = if trait_substs.needs_subst() {
```

Comment:
Why do we need to do this? Should this not be handled by `find_trait_method_substs`?